stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+    or nonempty, and in the latter case, to obtain its head and tail. *)
 
 and +'a node =
   | Nil
   | Cons of 'a * 'a t (**)
-(** A fully-evaluated list node, either empty or containing an element
-    and a delayed tail. *)
+(** A node is either [Nil], which means that the sequence is empty,
+    or [Cons (x, xs)], which means that [x] is the first element
+    of the sequence and that [xs] is the remainder of the sequence. *)
+
+(** {1 Consuming sequences} *)
+
+(**
+
+   The functions in this section consume their argument, a sequence, either
+   partially or completely:
+   - [is_empty], [head], [tail], [uncons] consume the sequence down to depth 1.
+     That is, they demand the first argument of the sequence, if there is one.
+   - [iter], [fold_left], [length], etc., consume the sequence all the way to
+     its end. The sequence must be finite; otherwise, they do not terminate.
+   - [for_all], [exists], [find], etc. consume the sequence down to a certain
+     depth, which is a priori unpredictable.
+
+   The functions that consume two sequences can be organized in two subgroups:
+   - [iter2], [fold_left2], [for_all2], etc., consume both sequences all the
+     way to the end, provided they have the same length. Otherwise, they fully
+     consume the shorter sequence, and ignore the remainder of the longer
+     sequence.
+   - [exists2], [equal], [compare] consume the sequences down to a certain
+     depth, which is a priori unpredictable.
+
+   None of the functions in this section is lazy. These functions
+   are consumers: they force some computation to take place. *)
+
+val is_empty:  'a t -> bool
+(** [is_empty xs] determines whether the sequence [xs] is empty.
+
+    @since 4.14 *)
+
+val head : 'a t -> 'a
+(** [head xs] is the first element of the sequence [xs].
+
+    The sequence [xs] must be nonempty.
+
+    @raise Invalid_argument if [xs] is empty.
+    @since 4.14 *)
+
+val tail : 'a t -> 'a t
+(** [tail xs] is the sequence [xs], deprived of its first element.
+
+    The sequence [xs] must be nonempty.
+
+    @raise Invalid_argument if [xs] is empty.
+    @since 4.14 *)
+
+val uncons : 'a t -> ('a * 'a t) option
+(** If [xs] is empty, then [uncons xs] is [None].
+
+    If [xs] is nonempty, then [uncons xs] is
+    [Some (head xs, tail xs)],
+    that is, a pair of the head and tail of the sequence [xs].
+
+    This equivalence holds if [xs] is persistent.
+    If [xs] is ephemeral, then [uncons] must be preferred
+    over separate calls to [head] and [tail],
+    which would cause [xs] to be queried twice.
+
+    @since 4.14 *)
+
+
+
+val length : 'a t -> int
+(** [length xs] is the length of the sequence [xs].
+
+    The sequence [xs] must be finite.
+
+    @since 4.14 *)
+
+val iter : ('a -> unit) -> 'a t -> unit
+(** [iter f xs] invokes [f x] successively
+    for every element [x] of the sequence [xs],
+    from left to right.
+
+    The sequence [xs] must be finite. *)
```

Comment:
> The sequence [xs] must be finite.

It seems valid to use this function (and the other traversals) with an exception-raising iterator on an infinite sequence.

Reply:
It also seems valid to use it on an infinite sequence, if we want an infinite loop (for instance something like `iter print_int primes`).